File: classlinesearch.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Automated GILDAS-CLASS Pipeline
-------------------------------
Line search mode
Version 1.4

Copyright (C) 2025 - Andrés Megías Toledano

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.
"""

# Libraries and functions.
import os
import gc
import argparse
import yaml
import numpy as np
import matplotlib.pyplot as plt
from astropy.io import fits
from scipy.stats import median_abs_deviation
from scipy.interpolate import UnivariateSpline


# Custom functions.

def full_path(text):
    """
    Obtain the full path described in a text string.
    
    Corrects the format of the path, allowing the use of operating system
    abbreviations (for example, './', '../' and '~' for Unix).
    
    Parameters
    ----------
    text : str
        Text of the path, which can include operating system abbreviations.
        
    Returns
    -------
    path : str
        Text of the full path, so that Python can use it.
    """
    path = text
    if path.startswith('~'):
        path = os.path.expanduser('~') + path[1:]  
    path = str(os.path.realpath(path))
    return path

def get_mask_from_windows(x, windows):
    """
    Select the regions of the input array specified by the given windows.

    Parameters
    ----------
    x : array
        Input data.
    windows : array
        Windows that specify the regions of the data.

    Returns
    -------
    mask : array (bool)
        Resultant mask array.
    """
    mask = np.ones(len(x), dtype=bool)
    for (x1, x2) in windows:
        mask *= (x < x1) + (x > x2)
    return mask

def rolling_function(func, y, size, **kwargs):
    """
    Apply a function in a rolling way, in windows of the specified size.

    Parameters
    ----------
    y : array
        Input data.
    func : function
        Function to be applied.
    size : int
        Size of the windows to group the data. It must be odd.
    **kwargs : (various)
        Keyword arguments of the function to be applied.

    Returns
    -------
    y_f : array
        Resultant array.
    """
    def rolling_window(y, window):
        """
        Group the input data according to the specified window size.
        
        Function by Erik Rigtorp.
        """
        shape = y.shape[:-1] + (y.shape[-1] - window + 1, window)
        strides = y.strides + (y.strides[-1],)
        y_w = np.lib.stride_tricks.as_strided(y, shape=shape, strides=strides)
        return y_w
    min_size = 1
    size = int(size) + (int(size) + 1) % 2
    size = max(min_size, size)
    N = len(y)
    y_c = func(rolling_window(y, size), -1, **kwargs)
    M = min(N, size) // 2
    y_1, y_2 = np.zeros(M), np.zeros(M)
    for i in range(M):
        j1 = 0
        j2 = max(min_size, 2*i)
        y_1[i] = func(y[j1:j2], **kwargs)
        j1 = N - max(min_size, 2*i)
        j2 = N
        y_2[-i-1] = func(y[j1:j2], **kwargs)
    y_f = np.concatenate((y_1, y_c, y_2))
    return y_f

def rolling_sigma_clip_mask(x, y, smooth_size, sigmas=6.0, iters=2):
    """
    Apply a rolling sigma clip and return a mask of the remaining data.

    Parameters
    ----------
    x : array
        Dependent variable.
    y : array
        Independent variable.
    size : int
        Size of the windows to group the data. It must be odd.
    sigmas : float, optional
        Number of standard deviations used as threshold. The default is 6.0.
    iters : int, optional
        Number of iterations performed. The default is 3.

    Returns
    -------
    mask : array (bool)
        Mask of the remaining data after applying the sigma clip.
    """
    mask = np.ones(len(y), dtype=bool)
    abs_y = abs(y)
    for i in range(iters):
        rolling_mad = rolling_function(median_abs_deviation, abs_y[mask],
                                       size=smooth_size, scale='normal')
        rolling_mad = np.interp(x, x[mask], rolling_mad)

        mask *= np.less(abs_y, sigmas*rolling_mad)
    return mask

def sigma_clip_mask(y, sigmas=6.0, iters=2):
    """
    Apply a sigma clip and return a mask of the remaining data.

    Parameters
    ----------
    y : array
        Input data.
    sigmas : float, optional
        Number of standard deviations used as threshold. The default is 4.0.
    iters : int, optional
        Number of iterations performed. The default is 3.

    Returns
    -------
    mask : array (bool)
        Mask of the remaining data after applying the sigma clip.
    """
    mask = np.ones(len(y), dtype=bool)
    abs_y = abs(y)
    for i in range(iters):
        mad = median_abs_deviation(abs_y[mask], scale='normal')
        mask *= abs_y < sigmas*mad
    return mask

def get_windows_from_mask(x, mask, margin=1., ref_width=8.):
    """
    Return the windows of the masked regions of the input array.

    Parameters
    ----------
    x : array
        Input data.
    mask : array (bool)
        Indices of the empty regions of data.
    margin : float, optional
        Relative margin added to the windows found initially.
        The default is 0.0.
    ref_width: float, optional
        Reference width of the windows.

    Returns
    -------
    windows : array (float)
        List of the inferior and superior limits of each window.
    inds : array (int)
        List of indices that define the filled regions if the data.
    """
    resolution = np.median(abs(np.diff(x)))
    all_inds = np.arange(len(x))
    diff_inds = np.diff(np.concatenate(([0], np.array(mask, dtype=int), [0])))
    cond1 = (diff_inds == 1)[:-1]  # from 0 to 1 
    cond2 = (diff_inds == -1)[1:]  # from 1 to 0 
    inds = np.append(all_inds[cond1], all_inds[cond2])
    inds = np.sort(inds).reshape(-1,2)
    windows = x[inds]
    for (i, window) in enumerate(windows):
        center = (window[0] + window[1]) / 2
        semiwidth = (window[1] - window[0]) / 2
        semiwidth = max(semiwidth, ref_width/4)
        semiwidth += margin*resolution
        windows[i,:] = [center - semiwidth, center + semiwidth]
    i = 0
    while i < len(windows):
        x1, x2 = windows[i]
        if (x2 - x1) > 6.*ref_width*resolution:
            windows = np.delete(windows, i, axis=0)
        else:
            i += 1
    if windows[0,0] <= x.min():
        windows = np.delete(windows, 0, axis=0)
    if windows[-1,1] >= x.max():
        windows = np.delete(windows, -1, axis=0)
    return windows

def fit_baseline(x, y, windows, smooth_size):
    """
    Fit the baseline of the curve ignoring the specified windows.

    Parameters
    ----------
    x : array
        Independent variable.
    y : array
        Dependent variable.
    windows : array
        Windows that specify the regions of the data.
    smooth_size : int
        Size of the filter applied for the fitting of the baseline.

    Returns
    -------
    y3 : array
        Baseline of the curve.
    """
    mask = get_mask_from_windows(x, windows)
    x_ = x[mask]
    y_ = y[mask]
    y_s = rolling_function(np.median, y_, smooth_size)
    s = sum((y_s - y_)**2)
    spl = UnivariateSpline(x_, y_, s=s)
    yf = spl(x)
    return yf

def identify_lines(x, y, smooth_size, line_width, sigmas, iters=2,
                   rolling_sigma_clip=False):
    """
    Identify the lines of the spectrum and fit the baseline.

    Parameters
    ----------
    x : array
        Frequency.
    y : array
        Intensity.
    smooth_size : int
        Size of the filter applied for the fitting of the baseline.
    line_width : float
        Reference line width for merging close windows.
    sigmas : float
        Threshold for identifying the outliers.
    iters : int, optional
        Number of iterations of the process. The default is 2.
    rolling_sigma_clip: bool, optional
        Use a rolling sigma clip for finding the outliers.

    Returns
    -------
    windows: array
        Values of the windows of the identified lines.
    """
    y_ = rolling_function(np.median, y, smooth_size)  
    for i in range(iters):
        if rolling_sigma_clip:
            mask = rolling_sigma_clip_mask(x, y-y_, 2*smooth_size, sigmas,
                                           iters=2)
        else:
            mask = sigma_clip_mask(y-y_, sigmas=sigmas, iters=2)
        windows = get_windows_from_mask(x, ~mask, margin=1.5, ref_width=line_width)
        if i+1 < iters:
            y_ = fit_baseline(x, y, windows, smooth_size)  
    return windows

def load_spectrum(file, load_fits=False):
    """
    Load the spectrum from the given input file.

    Parameters
    ----------
    file : str
        Path of the plain text file (.dat) to load, without the extension.
    load_fits : bool
        If True, load also a .fits file and return the HDU list. 

    Returns
    -------
    x : array
        Frequency.
    y : array
        Intensity.
    hdul : HDU list (astropy)
        List of the HDUs (Header Data Unit).
    """
    data = np.loadtxt(file + '.dat')
    x = data[:,0]
    y = data[:,1]
    if np.sum(np.isnan(data)) != 0:
        raise Exception('Data of file {} is corrupted.'.format(file))
    if load_fits:
        hdul = fits.open(file + '.fits')
    else:
        hdul = None
    return x, y, hdul
    

def save_yaml_dict(dictionary, file_path, default_flow_style=False, replace=False):
    """
    Save the input YAML dictionary into a file.

    Parameters
    ----------
    dictionary : dict
        Dictionary that wants to be saved.
    file_path : str
        Path of the output file.
    default_flow_style : bool, optional
        The flow style of the output YAML file. The default is False.
    replace : bool, optional
        If True, replace the output file in case it existed. If False, load the
        existing output file and merge it with the input dictionary.
        The default is False.

    Returns
    -------
    None.
    """
    file_path = os.path.realpath(file_path)
    if not replace and os.path.isfile(file_path):
        with open(file_path, 'r') as file:
            old_dict = yaml.safe_load(file)
        new_dict = {**old_dict, **dictionary}
    else:
        new_dict = dictionary
    with open(file_path, 'w') as file:
        yaml.dump(new_dict, file, default_flow_style=default_flow_style)

# Arguments.
parser = argparse.ArgumentParser()
parser.add_argument('folder', default='./')
parser.add_argument('file', default='spectrum.dat')
parser.add_argument('-smooth', default=20, type=int)
parser.add_argument('-width', default=6, type=int)
parser.add_argument('-threshold', default=6., type=float)
parser.add_argument('-plots_folder', default='plots', type=str)
parser.add_argument('--rolling_sigma', action='store_true')
parser.add_argument('--save_plots', action='store_true')
parser.add_argument('--check_windows', action='store_true')
args = parser.parse_args()
original_folder = full_path(os.getcwd())
os.chdir(full_path(args.folder))
np.seterr('raise')
    
#%% Calculations.

windows_dict = {}

# Processing of each spectrum.
for file in args.file.split(','):

    # Loading of the data file.
    frequency, intensity, _ = load_spectrum(file)

    # Identification of the lines and reduction of the spectrum.
    windows = identify_lines(frequency, intensity, smooth_size=args.smooth,
                       line_width=args.width, sigmas=args.threshold, iters=2,
                       rolling_sigma_clip=args.rolling_sigma)
    intensity_cont = fit_baseline(frequency, intensity, windows, args.smooth)
    intensity_red = intensity - intensity_cont
    
    # Windows.
    if len(windows) != 0:
        print('{} windows identified for {}.'
              .format(str(len(windows)), file))
    else:
        windows = np.array([[frequency[0], frequency[1]/2]])
        print('No lines identified for {}.'.format(file))
    windows_dict[file] = [[x1, x2] for x1, x2 in windows.tolist()]
    
    gc.collect()
        
    #%% Plots.
    
    if args.save_plots:
        
        y_lims = [np.quantile(intensity, 1e-3), np.quantile(intensity, 1.-1e-3)]
        y_lims = [y_lims[0] - np.diff(y_lims)/10, y_lims[1] + np.diff(y_lims)/10]
        yr_lims = [np.quantile(intensity_red, 5e-4), np.quantile(intensity_red, 1.-5e-4)]
        yr_lims = [yr_lims[0] - np.diff(yr_lims)/5, yr_lims[1] + np.diff(yr_lims)/5]
        
        fig = plt.figure(1, figsize=(10,7))
        plt.clf()
        plt.subplots_adjust(hspace=0, wspace=0)
        
        sp1 = plt.subplot(2,1,1)
        plt.step(frequency, intensity, where='mid', color='black', ms=6)
        for x1, x2 in windows:
            plt.axvspan(x1, x2, color='gray', alpha=0.3)
        plt.plot(frequency, intensity_cont, 'tab:green', label='fitted baseline')
        plt.locator_params(axis='both', nbins=10)
        plt.ticklabel_format(style='sci', useOffset=False)
        plt.margins(x=0)
        plt.ylim(y_lims)
        plt.xlabel('frequency (MHz)')
        plt.ylabel('original intensity (K)')
        plt.legend(loc='upper right')
        plt.title('Full spectrum - {}'.format(file), fontweight='semibold', pad=12.)
    
        plt.subplot(2,1,2, sharex=sp1)
        for (x1, x2) in windows:
            plt.axvspan(x1, x2, color='gray', alpha=0.3)
        plt.step(frequency, intensity_red, where='mid', color='black')
        plt.locator_params(axis='both', nbins=10)
        plt.ticklabel_format(style='sci', useOffset=False)
        plt.margins(x=0)
        plt.ylim(yr_lims)
        plt.xlabel('frequency (MHz)')
        plt.ylabel('reduced intensity (K)')

        fig.align_ylabels()
        plt.tight_layout(pad=0.7, h_pad=1.0)

        dx = np.median(np.diff(frequency)) 
        plt.rcParams['font.size'] = 8
        num_lines = len(windows)
        num_plots = 1 + (num_lines - 1)//15
        for i in range(num_plots):
            fig = plt.figure(2+i, figsize=(12, 6))
            plt.clf()
            for j in range(min(num_lines - 15*i, 15)):
                plt.subplot(3, 5, j+1)
                j += 15*i
                x1, x2 = windows[j]
                margin = max(args.width*dx, 0.4*(x2-x1))
                mask = (frequency > x1 - margin) * (frequency < x2 + margin)
                xj = frequency[mask]
                yrj = intensity_red[mask]
                plt.step(xj, yrj, where='mid', color='black')
                plt.axvspan(x1, x2, color='gray', alpha=0.2)
                plt.margins(x=0, y=0.1)
                plt.locator_params(axis='both', nbins=10)
                if j+1 > min(15*(i+1), num_lines) - 5:
                    plt.xlabel('frequency (MHz)')
                if j%5 == 0:
                    plt.ylabel('reduced intensity (K)')
                plt.xticks(fontsize=6)
                plt.yticks(fontsize=6)
                plt.locator_params(axis='x', nbins=1)
                plt.locator_params(axis='y', nbins=3)
                plt.ticklabel_format(style='sci', useOffset=False)
            window_num = ''
            if num_plots > 1:
                window_num = ' ({})'.format(i+1)
            plt.suptitle('Identified lines{} - {}'.format(window_num, file),
                             fontweight='semibold')
            fig.align_ylabels()
            plt.tight_layout(pad=1.2, h_pad=0.6, w_pad=0.1)
        
        if args.save_plots:
            os.chdir(original_folder)
            os.chdir(full_path(args.plots_folder))
            plt.figure(1)
            plt.savefig('spectrum-{}.png'.format(file))
            print("    Saved plot in {}spectrum-{}.png."
                  .format(args.plots_folder, file))
            for i in range(num_plots):
                plt.figure(2+i)
                plt.savefig('lines-{}_{}.png'.format(file, i+1))
                print("    Saved plot in {}lines-{}_{}.png."
                      .format(args.plots_folder, file, i+1))
            print()
            os.chdir(original_folder)
            os.chdir(full_path(args.folder))

        plt.close('all')
            
    gc.collect()
    
# Export of the frequency windows of each spectrum.
save_yaml_dict(windows_dict, 'frequency_windows.yaml', default_flow_style=None)
print('Saved windows in {}frequency_windows.yaml.'.format(args.folder))

print()

File: classlinesearch.jl
#!/usr/bin/env julia
"""
Automated GILDAS-CLASS Pipeline
-------------------------------
Line search mode
Version 1.3

Copyright (C) 2024 - Andrés Megías Toledano

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.
"""

# Libraries and functions.
import PyPlot as plt
import RollingFunctions: rolling
import Statistics as stats
import Random as random
import StatsBase: mad
import Interpolations: LinearInterpolation, Line
import SciPy.interpolate as scipy_interpolate
import Distributions: Normal
import Formatting: printfmtln
import ArgParse as argparse
import YAML as yaml
import FITSIO as fitsio
import DelimitedFiles as dlmfiles

# Custom functions.

function safe_realpath(path)
    """
    Return the full path corresponding to the input text.

    Parameters
    ----------
    path : String
        Text of the path to be converted.

    Returns
    -------
    full_path : String
        Text of the full path corresponding to the input text.
    """
    if ispath(path)
        return realpath(path)
    end
    if path != ""
        a, b = splitdir(path)
        return joinpath(safe_realpath(a), b)
    else
        return ""
    end
 end

function madn(x)
    """
    Return the normalized median absolute deviation (MAD) of the input data.

    Parameters
    ----------
    x : Array
        Input data.

    Returns
    -------
    y : Float
        Normalized median absolute deviation.
    """
    y = mad(x, normalize=true)
    return y
end

function rolling_function(func::Function, y::Vector{Float64}, roll_size::Int)
    """
    Apply a function in a rolling way, in windows of the specified size.

    Parameters
    ----------
    y : Vector
        Input data.
    func : Function
        Function to be applied.
    size : Int
        Size of the windows to group the data. It must be odd.

    Returns
    -------
    y_f : Vector
        Resultant array.
    """
    min_size = 1
    roll_size = Int(roll_size) + (Int(roll_size) + 1) % 2
    roll_size = max(min_size, roll_size)
    N = length(y)
    y_c = rolling(func, y, roll_size)
    M = min(N, roll_size) ÷ 2
    y_1 = zeros(M)
    y_2 = zeros(M)
    for i in 0:M-1
        j1 = 1
        j2 = max(min_size, 2*i)
        y_1[1+i] = func(y[j1:j2])
        j1 = N + 1 - max(min_size, 2*i)
        j2 = N
        y_2[end-i] = func(y[j1:j2])
    end
    y_f = vcat(y_1, y_c, y_2)
    return y_f
end

function get_windows(x::Vector{Float64}, cond::Vector{Bool};
                        margin::Float64=0.0, width::Float64=10.0)
    """
    Return the windows of the empty regions of the input array.

    Parameters
    ----------
    x : Vector
        Input data.
    cond : Vector
        Indices of the empty regions of data.
    margin : Float, optional
        Relative margin added to the windows found initially.
        The default is 0.0.
    width: Float, optional
        Minimum separation in points between two consecutive windows.

    Returns
    -------
    windows : Vector
        List of the inferior and superior limits of each window.
    inds : Vector
        List of indices that define the filled regions if the data.
    """

    N = length(x)
    separation = abs.(diff(x))
    reference = stats.median(separation)
    all_inds = 1:N

    var_inds = diff(vcat([0], Int.(cond), [0]))
    cond1 = (var_inds .== 1)[1:end-1]
    cond2 = (var_inds .== -1)[2:end]
    inds = vcat(all_inds[cond1], all_inds[cond2])
    inds = reshape(sort(inds), (2,length(inds)÷2))'

    windows = x[inds]
    for i in 1:size(windows)[1]
        window = windows[i,:]
        center = stats.mean(window)
        semiwidth = (window[2] - window[1]) / 2
        semiwidth = max(3*reference, semiwidth, 0.1*width*reference)
        semiwidth = (1 + margin) * semiwidth - 1E-9
        windows[i,:] = [center - semiwidth, center + semiwidth]
        windows[i,1] = max(x[1], windows[i,1])
        windows[i,2] = min(windows[i,2], x[end])
    end
    i = 1
    while i < size(windows)[1]
        difference = windows[i+1,1] - windows[i,2]
        if difference < width*reference
            windows[i,1] = min(windows[i,1], windows[i+1,1])
            windows[i,2] = max(windows[i,2], windows[i+1,2])
            windows = windows[1:end .!= i+1, :]
        else
            i += 1
        end
    end
    return windows
end

function regions_args(x::Vector{Float64}, windows::Matrix{Float64};
                      margin::Float64=0.0)
    """
    Select the regions of the input array specified by the given windows.

    Parameters
    ----------
    x : Vector
        Input data.
    wins : Matrix
        Windows that specify the regions of the data.

    Returns
    -------
    cond : Vector
        Resultant condition array.
    """
    cond = ones(Bool, length(x))
    dx = stats.median(diff(x))
    for i in 1:size(windows)[1]
        x1, x2 = windows[i,1], windows[i,2]
        cond .*= ((x .<= x1 - dx*margin) .+ (x .>= x2 + dx*margin))
    end
    return cond
end

function sigma_clip_args(y::Vector{Float64}; sigmas::Float64=6.0, iters::Int=2)
    """
    Apply a sigma clip and return a mask of the remaining data.

    Parameters
    ----------
    y : Vector
        Input data.
    sigmas : Float, optional
        Number of standard deviations used as threshold. The default is 6.0.
    iters : Int, optional
        Number of iterations performed. The default is 3.

    Returns
    -------
    cond : Vector
        Mask of the remaining data after applying the sigma clip.
    """
    cond = ones(Bool, length(y))
    abs_y = abs.(y)
    for i in 1:iters
        cond .*= abs_y .< sigmas*madn(abs_y[cond])
    end
    return cond
end

function rolling_sigma_clip_args(x::Vector{Float64}, y::Vector{Float64};
                                 smooth::Int, sigmas::Float64=6.0, iters::Int=2)
    """
    Apply a rolling sigma clip and return a mask of the remaining data.

    Parameters
    ----------
    x : Vector
        Dependent variable.
    y : Vector
        Independent variable.
    size : Int
        Size of the windows to group the data. It must be odd.
    sigmas : Float, optional
        Number of standard deviations used as threshold. The default is 4.0.
    iters : Int, optional
        Number of iterations performed. The default is 3.

    Returns
    -------
    cond : Vector
        Mask of the remaining data after applying the sigma clip.
    """
    cond = ones(Bool, length(y))
    abs_y = abs.(y)
    for i in 1:iters
        rolling_mad = rolling_function(madn, abs_y[cond], 2*smooth)
        itp = LinearInterpolation(x[cond], rolling_mad, extrapolation_bc=Line())
        rolling_mad = itp(x)
        cond .*= abs_y .< sigmas.*rolling_mad
    end
    return cond
end

function fit_baseline(x::Vector{Float64}, y::Vector{Float64};
                      windows::Matrix{Float64}, smooth_size::Int)
    """
    Fit the baseline of the curve ignoring the specified windows.

    Parameters
    ----------
    x : Vector
        Independent variable.
    y : Vector
        Dependent variable.
    windows : Matrix
        Windows that specify the regions of the data.
    smooth_size : Int
        Size of the filter applied for the fitting of the baseline.

    Returns
    -------
    yf : Vector
        Baseline of the curve.
    """
    cond = regions_args(x, windows)
    x_ = x[cond]
    y_ = y[cond]
    y_s = rolling_function(stats.median, y_, smooth_size)
    s = sum((y_s - y_).^2)
    spl = scipy_interpolate.UnivariateSpline(x_, y_, s=s)
    yf = spl(x)
    return yf
end

function identify_lines(x::Vector{Float64}, y::Vector{Float64}; smooth_size::Int,
    line_width::Float64, sigmas::Float64, iters::Int=2,
    rolling_sigma_clip::Bool=false)
    """
    Identify the lines of the spectrum and fits the baseline.

    Parameters
    ----------
    x : Vector
        Frequency.
    y : Vector
        Intensity.
    smooth_size : Int
        Size of the filter applied for the fitting of the baseline.
    line_width : Float
        Reference line width for merging close windows.
    sigmas : Float
        Threshold for identifying the outliers.
    iters : Int, optional
        Number of iterations of the process. The default is 2.
    rolling_sigma_clip: Bool, optional
        Use a rolling sigma clip for finding the outliers.

    Returns
    -------
    y3 : Vector
        Estimated baseline.
    windows: Matrix
        Values of the windows of the identified lines.
    """

    local y2 = rolling_function(stats.median, y, smooth_size)

    local windows

    for i in 1:iters

        cond = []
        if rolling_sigma_clip
            cond = rolling_sigma_clip_args(x, y.-y2, smooth=smooth_size,
                                            sigmas=sigmas)
        else
            cond = sigma_clip_args(y.-y2, sigmas=sigmas)
        end
        _cond = Vector(.!cond)

        windows = get_windows(x, _cond, margin=1.5, width=line_width)

        if i < iters
            y2 = fit_baseline(x, y, windows=windows, smooth_size=smooth_size)
        end

    end

    return windows
end

function load_spectrum(file::String; load_fits::Bool=false)
    """
    Load the spectrum from the given input file.

    Parameters
    ----------
    file : String
        Path of the plain text file (.dat) to load, without the extension.
    load_fits : Bool
        If true, load also a .fits file and return the HDU list.

    Returns
    -------
    x : Vector
        Frequency.
    y : Vector
        Intensity.
    hdul : HDU list (FITSIO)
        List of the HDUs (Header Data Unit).
    """
    data = dlmfiles.readdlm("$file.dat")
    x = data[:,1]
    y = data[:,2]
    if sum(isnan.(data)) != 0
        println("Data of file $file is corrupted.")
        throw(Exception)
    end
    if load_fits
        hdul = fitsio.FITS("$file.fits")
    else
        hdul = nothing
    end
    return x, y, hdul
end

function save_yaml_dict(dictionary::Dict, file_path::String; replace::Bool=false)
    """
    Save the input YAML dictionary into a file.

    Parameters
    ----------
    dictionary : Dict
        Dictionary that wants to be saved.
    file_path : String
        Path of the output file.
    default_flow_style : Bool, optional
        The flow style of the output YAML file. The default is False.
    replace : Bool, optional
        If true, replace the output file in case it existed. If false, load the
        existing output file and merge it with the input dictionary.
        The default is false.

    Returns
    -------
    nothing.
    """
    file_path = safe_realpath(file_path)
    if ! replace & isfile(file_path)
        old_dict = yaml.load_file(file_path)
        new_dict = merge(old_dict, dictionary)
    else
        new_dict = dictionary
    end
    yaml.write_file(file_path, new_dict)
end

# Arguments.
aps = argparse.ArgParseSettings()

argparse.@add_arg_table! aps begin
    "folder"
    arg_type = String
    required = true
    "file"
    arg_type = String
    required = true
    "--smooth"
    arg_type = Int
    default = 20
    "--width"
    arg_type = Float64
    default = 6.0
    "--threshold"
    arg_type = Float64
    default = 8.0
    "--plots_folder"
    arg_type = String
    default = "plots"
    "--rolling_sigma"
    action = :store_true
    "--save_plots"
    action = :store_true
end

#%%

args = argparse.parse_args(aps)
original_folder = safe_realpath(pwd())
cd(safe_realpath(args["folder"]))

#%% Calculations.

windows_dict = Dict()

# Processing of each spectrum.
for file in split(args["file"], ",")

    # Loading of the data file.
    frequency, intensity = load_spectrum(String(file))

    # Identification of the lines and reduction of the spectrum.
    windows =
        identify_lines(frequency, intensity, smooth_size=args["smooth"],
                       line_width=args["width"], sigmas=args["threshold"],
                       iters=2, rolling_sigma_clip=args["rolling_sigma"])
    intensity_cont = fit_baseline(frequency, intensity, windows=windows,
                                  smooth_size=args["smooth"])
    intensity_red = intensity .- intensity_cont

    # Windows.

    num_windows = size(windows)[1]

    if num_windows != 0
        println("$num_windows windows identified for $file.")
    else
        windows = [[frequency[1], frequency[2]/2]]
        println("No lines identified for $file.")
    end

    windows_dict[file] = [[windows[i,1], windows[i,2]] for i in 1:num_windows]

    #%% Plots.

    if args["save_plots"]

        plt.figure(1, figsize=(10,7))
        plt.clf()
        plt.subplots_adjust(hspace=0, wspace=0)

        sp1 = plt.subplot(2,1,1)
        plt.step(frequency, intensity, where="mid", color="black", ms=6)
        for i in 1:num_windows
            x1, x2 = windows[i,:]
            plt.axvspan(x1, x2, color="gray", alpha=0.3)
        end
        plt.plot(frequency, intensity_cont, "tab:green", label="fitted baseline")
        plt.ticklabel_format(style="sci", useOffset=false)
        plt.margins(x=0)
        plt.xlabel("frequency (MHz)")
        plt.ylabel("original intensity (K)")
        plt.legend(loc="upper right")

        plt.subplot(2,1,2, sharex=sp1)
        for i in 1:num_windows
            x1, x2 = windows[i,:]
            plt.axvspan(x1, x2, color="gray", alpha=0.3)
        end
        plt.step(frequency, intensity_red, where="mid", color="black")
        plt.ticklabel_format(style="sci", useOffset=false)
        plt.margins(x=0)
        plt.xlabel("frequency (MHz)")
        plt.ylabel("reduced intensity (K)")

        plt.suptitle("Full spectrum - $file", fontweight="semibold")
        plt.tight_layout(pad=0.7, h_pad=1.0)

        dx = stats.median(diff(frequency))
        plt.matplotlib.rc("font", size=8)
        num_lines = size(windows)[1]
        num_plots = 1 + (num_lines - 1)//15
        for i in 0:num_plots-1
            fig = plt.figure(2+i, figsize=(12, 6))
            plt.clf()
            for j in 0:Int(round(min(num_lines - 15*i, 15))-1)
                plt.subplot(3, 5, j+1)
                j += Int(15*i)
                x1, x2 = windows[j+1,:]
                margin = max(args["width"]*dx, 0.4*(x2-x1))
                cond = (frequency .> x1 - margin) .* (frequency .< x2 + margin)
                xj = frequency[cond]
                yrj = intensity_red[cond]
                plt.step(xj, yrj, where="mid", color="black")
                plt.axvspan(x1, x2, color="gray", alpha=0.2)
                plt.margins(x=0, y=0.1)
                if j+1 > min(15*(i+1), num_lines) - 5
                    plt.xlabel("frequency (MHz)")
                end
                if j%5 == 0
                    plt.ylabel("reduced intensity (K)")
                end
                plt.xticks(fontsize=6)
                plt.yticks(fontsize=6)
                plt.locator_params(axis="x", nbins=1)
                plt.locator_params(axis="y", nbins=3)
                plt.ticklabel_format(style="sci", useOffset=false)

            end
            window_num = ""
            if num_plots > 1
                window_num = " ($(Int(i)+1))"
            end
            plt.suptitle("Identified lines$window_num - $file",
                             fontweight="semibold")
            fig.align_ylabels()
            plt.tight_layout(pad=1.2, h_pad=0.6, w_pad=0.1)

        end

        if args["save_plots"]
            cd(original_folder)
            cd(safe_realpath(args["plots_folder"]))
            plt.figure(1)
            plt.savefig("spectrum-$file.png", dpi=200)
            print("    ")
            println("Saved plot in $(args["plots_folder"])spectrum-$file.png.")
            for i in 1:num_plots
                i = Int(i)
                plt.figure(1+i)
                plt.savefig("lines-$(file)_$i.png", dpi=200)
                print("    ")
                println("Saved plot in $(args["plots_folder"])lines-$(file)_$i.png.")
            end
            println()
            cd(original_folder)
            cd(safe_realpath(args["folder"]))
        end

    end

end

# Export of the frequency windows of each spectrum.
save_yaml_dict(windows_dict, "frequency_windows.yaml")
println("Saved windows in $(args["folder"])frequency_windows.yaml.")

println()
